lib/IDE/CursorInfo.cpp

```diff
@@ -242,6 +331,59 @@ class CursorInfoDoneParsingCallback : public IDEInspectionCallbacks {
     return CursorInfo;
   }
 
+  std::unique_ptr<ResolvedCursorInfo>
+  getExprResult(NodeFinderExprResult *ExprResult, SourceFile *SrcFile,
+                NodeFinder &Finder) const {
+    Expr *E = ExprResult->getExpr();
+    DeclContext *DC = ExprResult->getDeclContext();
+
+    // Type check the statemnt containing E and listen for solutions.
+    CursorInfoTypeCheckSolutionCallback Callback(E);
+    llvm::SaveAndRestore<TypeCheckCompletionCallback *> CompletionCollector(
+        DC->getASTContext().SolutionCallback, &Callback);
+    typeCheckASTNodeAtLoc(TypeCheckASTNodeAtLocContext::declContext(DC),
+                          E->getLoc());
+
+    if (Callback.getResults().empty()) {
+      // No results.
+      return nullptr;
+    }
+
+    for (auto Info : Callback.getResults()) {
+      // Type check the referenced decls so that all their parent closures are
+      // type-checked (see comment in typeCheckDeclAndParentClosures).
+      typeCheckDeclAndParentClosures(Info.ReferencedDecl);
+    }
+
+    if (Callback.getResults().size() != 1) {
+      // FIXME: We need to be able to report multiple results.
+      return nullptr;
+    }
+
+    // Deliver results
```

Comment:
All the results are reported using callback functions because they have pointers to stack variables – which they really shouldn’t have but that’s something else I want to fix. Thus, we cannot store the results, run the AST-based cursor info and then report them. Hence, what we do right now is serialize the results into a string and just deliver the AST-based results.

So: Delivering the results is a non-trivial change that I want to do in a follow-up PR, especially since we want to fall back to AST-based cursor info if solver-based cursor info doesn’t return any results.

Reply:
To be clear, we're now only delivering the one result right 😅?

Reply:
Yes, for now. Fixing the `TODO` in the line above is the next thing to do.

```cpp
// FIXME: We need to be able to report multiple results.
```